File: agent/internal/tool/git_show.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func gitShowTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "git_show",
			Description: "Show the details of a specific commit: diff, author, date, and message.",
			Category:    "git",
			Parameters: map[string]api.ParamDef{
				"path":   {Type: "string", Description: "Path to the git repository", Required: false, Default: "."},
				"commit": {Type: "string", Description: "Commit hash, branch name, or reference (e.g. HEAD~3)", Required: true},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			path, _ := params["path"].(string)
			if path == "" {
				path = "."
			}
			commit, _ := params["commit"].(string)
			if commit == "" {
				return nil, fmt.Errorf("commit is required")
			}
			out, err := gitCmd(ctx, path, "show", "--stat", "--pretty=fuller", commit)
			if err != nil {
				return nil, fmt.Errorf("git show failed: %w", err)
			}
			return map[string]any{"path": path, "commit": commit, "output": out}, nil
		},
	}
}

File: src/services/tools/git/gitShow.ts
import { z } from 'zod';
import type { ToolDef } from '../types';

export const gitShowTool: ToolDef = {
  name: 'git_show',
  description: 'Show the details of a specific commit, tag, or tree object. Returns full diff and metadata.',
  category: 'git',
  inputSchema: z.object({
    repoPath: z.string().describe('Absolute path to the git repository root.'),
    object: z.string().describe('The git object to show (e.g. commit hash "a1b2c3d", branch name "main", tag "v1.0").'),
  }),
  execute: async ({ repoPath, object }) => {
    const backendUrl = import.meta.env.VITE_API_URL || 'http://localhost:3001';
    try {
      const res = await fetch(`${backendUrl}/tools`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ tool: 'git_show', params: { repoPath, object } }),
      });
      if (!res.ok) throw new Error(`Backend error (${res.status})`);
      return res.json();
    } catch (error: any) {
      if (error.message?.includes('fetch')) {
        throw new Error('Git tools require the Express backend to be running (http://localhost:3001) or the Tauri desktop app');
      }
      throw error;
    }
  },
};
